store/s3storeKMS.go

```diff
@@ -0,0 +1,486 @@
+		ServerSideEncryption: aws.String(s3.ServerSideEncryptionAwsKms),
+		SSEKMSKeyId:          aws.String(s.KMSKey()),
+		Key:                  aws.String(path),
+		Body:                 bytes.NewReader(contents),
+	}
+
+	_, err := s.svc.PutObject(putObjectInput)
+	return err
+}
+
+func (s *S3KMSStore) readLatestFile(path string) (LatestIndexFile, error) {
+	getObjectInput := &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+
+	resp, err := s.svc.GetObject(getObjectInput)
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == "AccessDenied" {
+				// If we're not able to read the latest index for a KMS Key then proceed like it doesn't exist.
+				// We do this because in a chamber secret folder there might be other secrets written with a KMS Key that you don't have access to.
+				return LatestIndexFile{Latest: map[string]LatestValue{}}, nil
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				// Index doesn't exist yet, return an empty index
+				return LatestIndexFile{Latest: map[string]LatestValue{}}, nil
+			}
+		}
+		return LatestIndexFile{}, err
+	}
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return LatestIndexFile{}, err
+	}
+
+	var index LatestIndexFile
+	if err := json.Unmarshal(raw, &index); err != nil {
+		return LatestIndexFile{}, err
+	}
+
+	return index, nil
+}
+
+func (s *S3KMSStore) readLatest(service string) (LatestIndexFile, error) {
+	// Create an empty latest, this will be used to merge together the various KMS Latest Files
+	latestResult := LatestIndexFile{Latest: map[string]LatestValue{}}
+
+	// List all the files that are prefixed with kms and use them as latest.json files for that KMS Key.
+	params := &s3.ListObjectsInput{
+		Bucket: aws.String(s.bucket),
+		Prefix: aws.String(fmt.Sprintf("%s/__kms", service)),
+	}
+
+	var paginationError error
+
+	err := s.svc.ListObjectsPages(params, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+		for index := range page.Contents {
+			key_name := *page.Contents[index].Key
+			result, err := s.readLatestFile(key_name)
+
+			if err != nil {
+				paginationError = errors.New(fmt.Sprintf("Error reading latest index for KMS Key (%s): %s", key_name, err))
+				return false;
+			}
+
+			// Check if the chamber key already exists in the index.Latest map.
+			// Prefer the most recent version.
+			for k, v := range result.Latest {
+				if val, ok := latestResult.Latest[k]; ok {
+					if val.Version > v.Version {
+						latestResult.Latest[k] = val
+					} else {
+						latestResult.Latest[k] = v
+					}
+				} else {
+					latestResult.Latest[k] = v
+				}
+			}
+		}
+
+		return !lastPage
+	})
+
+	if paginationError != nil {
+		return latestResult, paginationError
+	}
+
+	if err != nil {
+		return latestResult, err
+	}
+
+	return latestResult, nil
+}
+
+func (s *S3KMSStore) latestFileKeyNameByKMSKey() string {
+	return fmt.Sprintf("__kms_%s__latest.json", strings.Replace(s.KMSKey(), "/", "_", -1))
```

Comment:
`KMSKey()` could be an alias (and usually is). What happens when you have multiple aliases pointing to the same key, and you use each of them? You'd be using different `latest` files depending on which alias you used.

Reply:
Yeah, we would need to be able to resolve the alias to the KMS Key. It's totally fine to do that but it just adds more permissions required for using this backend.

Instead, we could read the old KMS `latest` file, delete the key from that file, write the key to the new alias file and then write both files using their respective keys.

Reply:
Ok, yeah, I was just thinking out loud. It's probably best to assume they're separate keys and not resolve aliases. This might mean some superfluous writes or extra `latest`s, but meh.